ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala

```diff
@@ -705,40 +677,24 @@ private object OracleQueries extends Queries {
             fr" OR ",
           )
       }
-      import Queries.CompatImplicits.catsReducibleFromFoldable1
-      val outerSelectList =
-        sql"""contract_id, template_id, key, payload, agreement_text,
-              pctSignatories, pctObservers"""
-      // % is explicitly reserved by specification as a delimiter
-      val dupQ =
-        sql"""SELECT c.contract_id contract_id, $tpid template_id, key, payload, agreement_text,
-                     sd.parties pctSignatories, od.parties pctObservers,
-                     row_number() over (PARTITION BY c.contract_id ORDER BY c.contract_id) AS rownumber
-              FROM (contract c
-                    LEFT JOIN signatories sm ON (c.contract_id = sm.contract_id)
-                    LEFT JOIN observers om ON (c.contract_id = om.contract_id))
-                   LEFT JOIN (SELECT contract_id, LISTAGG(party, '%') parties
-                              FROM signatories GROUP BY contract_id) sd
-                          ON (c.contract_id = sd.contract_id)
-                   LEFT JOIN (SELECT contract_id, LISTAGG(party, '%') parties
-                              FROM observers GROUP BY contract_id) od
-                          ON (c.contract_id = od.contract_id)
-              WHERE (${Fragments.in(fr"sm.party", parties)}
-                     OR ${Fragments.in(fr"om.party", parties)})
-                    AND $queriesCondition"""
-      // see https://github.com/digital-asset/daml/issues/9388#issuecomment-820538688
-      // for a demonstration
-      val q = sql"SELECT $outerSelectList FROM ($dupQ) WHERE rownumber = 1"
+      // Oracle SQL/JSON path expression syntax: https://docs.oracle.com/en/database/oracle/oracle-database/19/adjsn/json-path-expressions.html#GUID-AEBAD813-99AB-418A-93AB-F96BC1658618
+      val quotedParties = parties.toVector.map(p => s""""$p"""").mkString(", ")
+      val q =
+        sql"""SELECT c.contract_id contract_id, $tpid template_id, key, payload, signatories, observers, agreement_text
+                FROM contract c
+                WHERE (json_exists(signatories, '$$[*]?(@ in ($quotedParties))')
+                       OR json_exists(observers, '$$[*]?(@ in ($quotedParties))')
```

Comment:
The key question for this PR: can we index the database such that these conditions are not applied linearly?

Also I strongly suggest two modifications to your path construction:

1. Put them together with `Cord` and pass the whole path through `oracleShortPathEscape`.
2. If possible, pass `quotedParties` using the `PASSING ? AS X` binding syntax used elsewhere in this file.

Reply:
Regarding the changes to the code: I've addressed the first point in https://github.com/digital-asset/daml/pull/9484/commits/b285faee00cb844cc44e9d26f689f85747106e22, as for the second I've done a few experiments but nothing worked and I always ended up with something which didn't work. I believe I got close to something working but I'm not sure of how to interpolate the parties vector as a list of values (e.g. `("Alice", "Bob")`).

Reply:
As for the main point, I'm doing a bit of experimentation, but I believe the [JSON Search Index](https://docs.oracle.com/en/database/oracle/oracle-database/19/adjsn/indexes-for-json-data.html#GUID-D7A604E1-F617-4C92-A9D9-00B6AB78EFD8) you brought up should do the job.

Reply:
> I believe I got close to something working but I'm not sure of how to interpolate the parties vector as a list of values

The trouble as it stands is that if the sum of lengths of party IDs in the request is >4000 bytes or so, you simply won't be able to query.

Reply:
A JSON Search Index gets the job done, I tested it on a table with 10K items, each having from 1 to 10 parties in a JSON CLOB column. The query sped up significantly and the execution plan reflected the usage of the JSON Search Index I created.